File: python/1930.py
# Link: https://leetcode.com/problems/unique-length-3-palindromic-subsequences/description/
# Runtime: 549 ms
# Memory: 18.12 MB

class Solution:
    def countPalindromicSubsequence(self, s: str) -> int:
        first_occurrance_index, last_occurrance_index = [-1] * 26, [-1] * 26
        for index, symbol in enumerate(s):
            if first_occurrance_index[ord(symbol) - ord('a')] == -1:
                first_occurrance_index[ord(symbol) - ord('a')] = index
            
            last_occurrance_index[ord(symbol) - ord('a')] = index
        
        number_of_palindromes_length3 = 0
        for i in range(26):
            if first_occurrance_index[i] != last_occurrance_index[i]:
                chars_between = set()
                for j in range(first_occurrance_index[i] + 1, last_occurrance_index[i]):
                    chars_between.add(s[j])

                number_of_palindromes_length3 += len(chars_between)
        
        return number_of_palindromes_length3

File: c++/1930.cpp
// Link: https://leetcode.com/problems/unique-length-3-palindromic-subsequences/description/
// Runtime: 131 ms
// Memory: 14.65 MB

class Solution 
{
public:
    int countPalindromicSubsequence(const std::string& s) 
    {
        std::vector<int> firstOccurrenceIndex(26, -1);
        std::vector<int> lastOccurrenceIndex(26, -1);
        for (int i = 0; i < static_cast<int>(s.size()); i++) 
        {
            if (firstOccurrenceIndex[s[i] - 'a'] == -1)
                firstOccurrenceIndex[s[i] - 'a'] = i;
            
            lastOccurrenceIndex[s[i] - 'a'] = i;
        }

        int numberOfPalindromesLength3 = 0;
        for (int i = 0; i < 26; i++)
            if (firstOccurrenceIndex[i] != lastOccurrenceIndex[i])
            {
                std::unordered_set<char> charsBetween;
                for (int j = firstOccurrenceIndex[i] + 1; j < lastOccurrenceIndex[i]; j++)
                    charsBetween.insert(s[j]);
                
                numberOfPalindromesLength3 += static_cast<int>(charsBetween.size());
            }

        return numberOfPalindromesLength3;
    }
};
